docs/implementing-cards.md

```diff
@@ -1,388 +1,570 @@
+### Replacement effects
+Some abilities allow the player to cancel or modify an effect. These abilities are always defined with the word "instead" or "would." Some examples:
+- Shield, which cancels the normal resolution of damage and replaces it with another effect (defeating the shield token)
+- Boba Fett's armor, which modifies the normal resolution of an instance of damage and reduces its value by 2
 
-Generally, it's a good idea to pass at least a `cardType` property, as that will automatically change the prompt to make it easier for the player to understand what is going on. Most other properties that apply to `Game.promptForSelect` are also valid here.
+These abilities are called "replacement effects" in the SWU rules and are defined using the `addReplacementEffectAbility` method. Otherwise the ability is defined very similar to a triggered ability, except that it has a `replaceWith` property object which defines an optional replacement effect in the `replacementImmediateEffect` sub-property. If `replacementImmediateEffect` is null, the triggering effect is canceled with no replacement. An optional `target` sub-property is also availabe to define a target for the replacement effect.
 
+Here is the Shield implementation as an example:
 ```javascript
-this.action({
-    title: 'Grant Covert to a character',
-    target: {
-        cardType: 'character',
-        location: 'play area'
+this.addReplacementEffectAbility({
+    title: 'Defeat shield to prevent attached unit from taking damage',
+    when: {
+        onDamageDealt: (event, context) => event.card === (context.source as UpgradeCard).parentCard
     },
-    // ...
+    replaceWith: {
+        target: this,
+        replacementImmediateEffect: AbilityHelper.immediateEffects.defeat()
+    },
+    effect: 'shield prevents {1} from taking damage',
+    effectArgs: (context) => [(context.source as UpgradeCard).parentCard],
 });
 ```
 
+## Ability building blocks
+This section describes some of the major components that are used in the definitions of abilities:
+- Context objects
+- Game systems
+- Target resolvers
+
+### Context object
+
+When the game starts to resolve an ability, it creates a context object for that ability. Generally, the context ability has the following structure:
+
 ```javascript
-this.action({
-    title: 'Sacrifice to discard an attachment'
-    target: {
-        cardType: 'attachment',
-        gameAction: AbilityDsl.actions.discardFromPlay()
-    },
-    // ...
+class AbilityContext {
+    constructor(properties) {
```

Comment:
Add a TODO to make this look more SWU-ish (this could apply in a few places)

Reply:
Can you elaborate? What would make the context class more SWUish?

Reply:
@AMMayberry1 Stuff like rings and stage being removed for example. It isn't high priority, but someone might assume this is an accurate breakdown of the context object when they're new.

Reply:
Oh lol right I don't even notice ring stuff anymore, good catch I'll remove it.

I think Stage might actually still be relevant (I think it means "ability resolution stage") but we're not using it in many places. I'll add a TODO in the code for later investigation to see if we still need it.